clique-grpc/src/lib.rs

```diff
@@ -1,18 +1,79 @@
+use futures::{Async, Future, Poll, Stream};
+use http::Uri;
+use hyper::client::connect::{Destination, HttpConnector};
+use tower_grpc::Request;
+use tower_hyper::{
+    client::{self, Connection},
+    util,
+};
+use tower_request_modifier;
+use tower_service::Service;
+use tower_util;
+use tower_util::MakeService;
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
 pub mod clique_proto {
     include!(concat!(env!("OUT_DIR"), "/messaging.rs"));
 }
 
-pub use self::clique_proto::*;
+pub struct UniquePool<S, T, R>
+where
+    S: MakeService<T, R> + Clone,
+    T: Hash + Eq,
+{
+    pool: HashMap<T, S::Service>,
+    maker: S,
+}
 
-use clique::messaging::{Broadcast, RapidResponseFuture};
+impl<S, T, R> UniquePool<S, T, R>
+where
+    S: MakeService<T, R> + Clone,
+    T: Hash + Eq,
+{
+    pub fn new(maker: S) -> UniquePool<S, T, R> {
+        UniquePool {
+            pool: HashMap::new(),
+            maker,
+        }
+    }
+}
 
-struct CliqueClient;
+trait Key<T> {
+    fn get_key(&self) -> T;
+}
+
+impl<M, T, R> Service<R> for UniquePool<M, T, R>
+where
+    M: MakeService<T, R> + Clone,
+    T: Hash + Eq,
+    R: Key<T>,
+    M::Service: Clone,
+{
+    type Response = M::Response;
+    type Error = (); // Need better error handling
+    type Future = Box<dyn Future<Item = Self::Response, Error = Self::Error>>;
 
-impl Broadcast for CliqueClient {
-    fn broadcast(&self, req: RapidRequest) -> RapidResponseFuture {
-        unimplemented!()
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        Ok(Async::Ready(()))
     }
-    fn set_membership(&self, reciepients: Vec<Endpoint>) {
-        unimplemented!()
+
+    fn call(&mut self, req: R) -> Self::Future {
+        if let Some(svc) = self.pool.get(&req.get_key()) {
+            Box::new(svc.call(req).map_err(|_| ()))
+        } else {
+            Box::new(
```

Comment:
small nit but style wise I generally prefer to put the future in a `let fut = self.maker...`
then i do `Box::new(fut)`, I find it makes it a bit easier to read.